#1 Importing Breast Cancer Dataset from Kaggle and Loading it

In [6]:
import pandas as pd
import seaborn as sns
import os
os.environ['KAGGLE_USERNAME'] = 'abirkrishnabanerjee'
os.environ['KAGGLE_KEY'] = '7a3202345159554ca809d4101b90266a'

In [3]:
! kaggle datasets download -d uciml/breast-cancer-wisconsin-data

Dataset URL: https://www.kaggle.com/datasets/uciml/breast-cancer-wisconsin-data
License(s): CC-BY-NC-SA-4.0
  0% 0.00/48.6k [00:00<?, ?B/s]
100% 48.6k/48.6k [00:00<00:00, 127MB/s]


In [16]:
! unzip /content/breast-cancer-wisconsin-data.zip

Archive:  /content/breast-cancer-wisconsin-data.zip
replace data.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
  inflating: data.csv                


#2 Removing Null values from Dataset

In [17]:
df = pd.read_csv('/content/data.csv')

In [18]:
df.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

In [19]:
df.shape

(569, 33)

In [20]:
df.isnull().sum()

id                           0
diagnosis                    0
radius_mean                  0
texture_mean                 0
perimeter_mean               0
area_mean                    0
smoothness_mean              0
compactness_mean             0
concavity_mean               0
concave points_mean          0
symmetry_mean                0
fractal_dimension_mean       0
radius_se                    0
texture_se                   0
perimeter_se                 0
area_se                      0
smoothness_se                0
compactness_se               0
concavity_se                 0
concave points_se            0
symmetry_se                  0
fractal_dimension_se         0
radius_worst                 0
texture_worst                0
perimeter_worst              0
area_worst                   0
smoothness_worst             0
compactness_worst            0
concavity_worst              0
concave points_worst         0
symmetry_worst               0
fractal_dimension_worst      0
Unnamed: 32                569
dtype: int64

In [21]:
df = df.drop('Unnamed: 32', axis=1)

In [22]:
df.shape

(569, 32)

In [23]:
df['diagnosis'].value_counts()

diagnosis
B    357
M    212
Name: count, dtype: int64

#3 Label Encoding M as 1 and B as 0

In [24]:
df.dtypes

id                           int64
diagnosis                   object
radius_mean                float64
texture_mean               float64
perimeter_mean             float64
area_mean                  float64
smoothness_mean            float64
compactness_mean           float64
concavity_mean             float64
concave points_mean        float64
symmetry_mean              float64
fractal_dimension_mean     float64
radius_se                  float64
texture_se                 float64
perimeter_se               float64
area_se                    float64
smoothness_se              float64
compactness_se             float64
concavity_se               float64
concave points_se          float64
symmetry_se                float64
fractal_dimension_se       float64
radius_worst               float64
texture_worst              float64
perimeter_worst            float64
area_worst                 float64
smoothness_worst           float64
compactness_worst          float64
concavity_worst            float64
concave points_worst       float64
symmetry_worst             float64
fractal_dimension_worst    float64
dtype: object

In [25]:
from sklearn.preprocessing import LabelEncoder
labelencoder = LabelEncoder()
df.iloc[:, 1]=labelencoder.fit_transform(df.iloc[:, 1].values)

In [26]:
df

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0      842302         1        17.99         10.38          122.80     1001.0   
1      842517         1        20.57         17.77          132.90     1326.0   
2    84300903         1        19.69         21.25          130.00     1203.0   
3    84348301         1        11.42         20.38           77.58      386.1   
4    84358402         1        20.29         14.34          135.10     1297.0   
..        ...       ...          ...           ...             ...        ...   
564    926424         1        21.56         22.39          142.00     1479.0   
565    926682         1        20.13         28.25          131.20     1261.0   
566    926954         1        16.60         28.08          108.30      858.1   
567    927241         1        20.60         29.33          140.10     1265.0   
568     92751         0         7.76         24.54           47.92      181.0   

     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0            0.11840           0.27760         0.30010              0.14710   
1            0.08474           0.07864         0.08690              0.07017   
2            0.10960           0.15990         0.19740              0.12790   
3            0.14250           0.28390         0.24140              0.10520   
4            0.10030           0.13280         0.19800              0.10430   
..               ...               ...             ...                  ...   
564          0.11100           0.11590         0.24390              0.13890   
565          0.09780           0.10340         0.14400              0.09791   
566          0.08455           0.10230         0.09251              0.05302   
567          0.11780           0.27700         0.35140              0.15200   
568          0.05263           0.04362         0.00000              0.00000   

     ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0    ...        25.380          17.33           184.60      2019.0   
1    ...        24.990          23.41           158.80      1956.0   
2    ...        23.570          25.53           152.50      1709.0   
3    ...        14.910          26.50            98.87       567.7   
4    ...        22.540          16.67           152.20      1575.0   
..   ...           ...            ...              ...         ...   
564  ...        25.450          26.40           166.10      2027.0   
565  ...        23.690          38.25           155.00      1731.0   
566  ...        18.980          34.12           126.70      1124.0   
567  ...        25.740          39.42           184.60      1821.0   
568  ...         9.456          30.37            59.16       268.6   

     smoothness_worst  compactness_worst  concavity_worst  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     concave points_worst  symmetry_worst  fractal_dimension_worst  
0                  0.2654          0.4601                  0.11890  
1                  0.1860          0.2750                  0.08902  
2                  0.2430          0.3613                  0.08758  
3                  0.2575          0.6638                  0.17300  
4                  0.1625          0.2364                  0.07678  
..                    ...             ...                      ...  
564                0.2216       

#4 Performing Linear Regression on the dataset

# Task
Perform linear regression on the breast cancer dataset to predict a suitable target variable using the available features. Calculate and report the R-squared value and other relevant metrics.

## Prepare the data

### Subtask:
Separate the features (independent variables) from the target variable (dependent variable).


**Reasoning**:
Separate the features (independent variables) from the target variable (dependent variable) as instructed.



In [27]:
X = df.drop(['diagnosis', 'id'], axis=1)
y = df['diagnosis']

## Split the data

### Subtask:
Split the dataset into training and testing sets.


**Reasoning**:
Split the dataset into training and testing sets for model training and evaluation.



In [28]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## Train the model

### Subtask:
Train a Linear Regression model on the training data.


**Reasoning**:
Train a Linear Regression model on the training data.



In [29]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

## Evaluate the model

### Subtask:
Make predictions on the testing data and calculate evaluation metrics like R-squared.


**Reasoning**:
Make predictions on the test data and calculate the R-squared score to evaluate the model's performance.



In [30]:
y_pred = model.predict(X_test)
from sklearn.metrics import r2_score
r2 = r2_score(y_test, y_pred)
print(f'R-squared: {r2}')

R-squared: 0.727101612622355


## Summary:

### Data Analysis Key Findings

*   The R-squared value for the linear regression model predicting the diagnosis is approximately 0.727. This indicates that about 72.7% of the variance in the diagnosis can be explained by the features used in the model.

### Insights or Next Steps

*   While the R-squared value of 0.727 suggests a reasonably good fit for a linear model on this dataset, it's important to note that the target variable 'diagnosis' is binary (likely representing malignant or benign). Linear regression is typically used for continuous target variables.
*   Given the nature of the target variable, a classification model (e.g., Logistic Regression, Support Vector Machine, Random Forest) would be a more appropriate choice for this problem. Evaluating the performance of classification models using relevant metrics (like accuracy, precision, recall, F1-score, and AUC) would provide a more suitable assessment of the model's ability to predict breast cancer diagnosis.
